File: leetcode-04/solutions/cpp/0567.cpp
class Solution {
 public:
  bool checkInclusion(string s1, string s2) {
    const int m = s1.length();
    const int n = s2.length();

    vector<int> map1(26);
    vector<int> map2(26);

    for (char c : s1) map1[c - 'a']++;

    for (int i = 0; i < n; i++) {
      if (i >= m) map2[s2[i - m] - 'a']--;
      map2[s2[i] - 'a']++;
      if (map1 == map2) return true;
    }

    return false;
  }
};

File: leetcode-04/solutions/python3/0567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        m = len(s1)
        n = len(s2)

        map1 = [0] * 26
        map2 = [0] * 26

        for c in s1:
            map1[ord(c) - ord('a')] += 1

        for i in range(n):
            if i >= m:
                map2[ord(s2[i - m]) - ord('a')] -= 1
            map2[ord(s2[i]) - ord('a')] += 1
            if map1 == map2:
                return True

        return False
